common/autoware_overlay_rviz_plugin/autoware_minimap_overlay_rviz_plugin/src/goal_pose.cpp

```diff
@@ -0,0 +1,158 @@
+{
+  std::string package_path =
+    ament_index_cpp::get_package_share_directory("autoware_minimap_overlay_rviz_plugin");
+  std::string goal_image_path = package_path + "/icons/goal.png";
+  goal_image_ = QImage(goal_image_path.c_str());
+  goal_image_ = goal_image_.scaled(20, 20, Qt::KeepAspectRatio);
+}
+
+void GoalPose::setGoalPosition(double local_x, double local_y, double origin_lat, double origin_lon)
+{
+  std::tie(goal_lat_, goal_lon_) = localToGeographic(local_x, local_y, origin_lat, origin_lon);
+}
+
+void GoalPose::setProjectionInfo(const std::string & projector_type, const std::string & mgrs_grid)
+{
+  projector_type_ = projector_type;
+  mgrs_grid_ = mgrs_grid;
+}
+
+std::pair<double, double> GoalPose::localToGeographic(
+  double local_x, double local_y, double origin_lat, double origin_lon)
+{
+  if (projector_type_ == tier4_map_msgs::msg::MapProjectorInfo::MGRS) {
+    return localToGeographicMGRS(local_x, local_y, origin_lat, origin_lon, mgrs_grid_);
+  } else if (projector_type_ == tier4_map_msgs::msg::MapProjectorInfo::LOCAL_CARTESIAN_UTM) {
+    return localToGeographicUTM(local_x, local_y, origin_lat, origin_lon);
+  }
+  // Return a default pair if none of the conditions match
+  return std::make_pair(0.0, 0.0);  // Adjust as necessary
+}
+
+std::pair<double, double> GoalPose::localToGeographicUTM(
+  double local_x, double local_y, double origin_lat, double origin_lon)
+{
+  int zone;
+  bool north_p;
+  double origin_x, origin_y, gamma, k;
+
+  // Convert origin coordinates to UTM
+  GeographicLib::UTMUPS::Forward(
+    origin_lat, origin_lon, zone, north_p, origin_x, origin_y, gamma, k);
+
+  // Add local coordinates to origin UTM coordinates
+  double goal_x = origin_x + local_x;
+  double goal_y = origin_y + local_y;
+
+  // Convert UTM coordinates back to geographic coordinates
+  double goal_lat, goal_lon;
+  GeographicLib::UTMUPS::Reverse(zone, north_p, goal_x, goal_y, goal_lat, goal_lon);
+
+  return {goal_lat, goal_lon};
+}
+
+std::pair<double, double> GoalPose::localToGeographicMGRS(
+  double local_x, double local_y, double origin_lat, double origin_lon,
+  const std::string & mgrs_grid)
+{
+  int zone;
+  bool north_p;
+  double origin_x, origin_y, gamma, k, mgrs_x, mgrs_y;
+
+  // Convert origin coordinates to UTM
+  GeographicLib::UTMUPS::Forward(
+    origin_lat, origin_lon, zone, north_p, origin_x, origin_y, gamma, k);
+
+  // Convert MGRS grid string to UTM coordinates
+  std::string mgrs_string = mgrs_grid + "0000000000000000";
+  int prec;
+  GeographicLib::MGRS::Reverse(mgrs_string, zone, north_p, mgrs_x, mgrs_y, prec);
+
+  // Calculate the global UTM coordinates by adding local offsets
+  double global_x = mgrs_x + local_x;
+  double global_y = mgrs_y + local_y;
+
+  // Convert global UTM coordinates to geographic coordinates
+  double goal_lat, goal_lon;
+  GeographicLib::UTMUPS::Reverse(zone, north_p, global_x, global_y, goal_lat, goal_lon);
+
+  return {goal_lat, goal_lon};
+}
```

Comment:
I think we can simplify this function.
(reduce the number of arguments, remove unnecessary call of `GegraphicLib::UTMUPS::Forward()`, ect)

Note that you would have to update the function definition in the header file as well.

```suggestion
std::pair<double, double> GoalPose::localToGeographicMGRS(
  double local_x, double local_y, const std::string & mgrs_grid)
{
  int zone;
  bool north_p;
  double mgrs_origin_x, mgrs_origin_y;
  int prec;

  // Convert MGRS grid string to UTM coordinates
  GeographicLib::MGRS::Reverse(mgrs_grid, zone, north_p, mgrs_origin_x, mgrs_origin_y, prec, false);

  // Calculate the global UTM coordinates by adding local offsets
  double utm_x = mgrs_origin_x + local_x;
  double utm_y = mgrs_origin_y + local_y;

  // Convert global UTM coordinates to geographic coordinates
  double goal_lat, goal_lon;
  GeographicLib::UTMUPS::Reverse(zone, north_p, utm_x, utm_y, goal_lat, goal_lon);

  return {goal_lat, goal_lon};
}
```